tests/realm.cpp

```diff
@@ -383,3 +383,95 @@ TEST_CASE("ShareRealm: in-memory mode from buffer") {
         REQUIRE_THROWS(Realm::get_shared_realm(config3)); // both buffer and encryption
     }
 }
+
+TEST_CASE("ShareRealm: realm closed in notify callback") {
+    TestFile config;
+    config.schema_version = 1;
+    config.schema = Schema{
+        {"object", {
+            {"value", PropertyType::Int, "", "", false, false, false}
+        }},
+    };
+    config.cache = false;
+    auto r1 = Realm::get_shared_realm(config);
+
+    r1->begin_transaction();
+    auto table = r1->read_group().get_table("class_object");
+    auto row_idx = table->add_empty_row(1);
+    auto table_idx = table->get_index_in_group();
+    r1->commit_transaction();
+
+    struct TestContext : public BindingContext {
+        SharedRealm& shared_realm;
+        TestContext(SharedRealm& r) : shared_realm(r) { }
+        //virtual void changes_available() = 0;
+    };
+
+    SECTION("changes_available") {
+        struct Context : public TestContext {
+            Context(SharedRealm& r) : TestContext(r) { }
+            virtual void changes_available() override
+            {
+                shared_realm->close();
+                shared_realm.reset();
+            }
+
+            virtual void will_change(std::vector<ObserverState> const&, std::vector<void*> const&) override
+            {
+                REQUIRE(false);
+            }
+
+            virtual void did_change(std::vector<ObserverState> const&, std::vector<void*> const&, bool) override
+            {
+                REQUIRE(false);
+            }
+        };
+        r1->m_binding_context.reset(new Context(r1));
+    }
+
+    SECTION("will_change") {
```

Comment:
Yes, will_change() and did_change() are still needed and their API and semantics are unchanged. Not handling the Realm being closed from within them is perfectly acceptable.